### PR TITLE
add(rpc): note tree sizes to `getblock` api

### LINKED ISSUE #6952: Add note tree size fields to the `getblock` RPC
## Motivation

Be compatible with the new `zcashd` RPC fields used by `lightwalletd`.

### Specifications

- New fields are in https://github.com/zcash/zcash/pull/6677/files#diff-decae4be02fb8a47ab4557fe74a9cb853bdfa3ec0fa1b515c0a1e5de91f4ad0bR781-R787

```
"trees": {                 (object) information about the note commitment trees
      "sapling": {             (object, optional)
          "size": n,             (numeric) the total number of Sapling note commitments as of the end of this block
      },
      "orchard": {             (object, optional)
          "size": n,             (numeric) the total number of Orchard note commitments as of the end of this block
      },
},
```

### Complex Code or Requirements

State lookups are concurrent, so we'll need to deal with concurrent database updates where the database overlaps the cloned non-finalized state.

Zebra's RPCs must remain compatible with the latest version of `lightwalletd`. If fields or methods are provided by `zcashd` but not used by `lightwalletd`, Zebra doesn't need to add them.

### Implementation Notes

I think we can implement this using our existing trees, [bridgetree::NonEmptyFrontier::position()](https://docs.rs/bridgetree/latest/bridgetree/struct.NonEmptyFrontier.html#method.position), and `u64 from Position` or `serde::Serialize`

`zcashd` implemented these RPC changes in https://github.com/zcash/zcash/pull/6677

### Testing

New state fields will need:
- [ ] the same tests as the RPCs, if they use a populated state

New fields in existing RPC methods will need:
- [ ] extra checks in existing unit tests
- [ ] snapshot updates (automated, see the snapshot test module for details)
- [ ] extra checks in existing integration tests
- [ ] manual `zcash-rpc-diff` test results in a comment on the PR, to make sure we match `zcashd`

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

We need to add sapling and orchard tree sizes to `getblock` method. 

Close https://github.com/ZcashFoundation/zebra/issues/6952 

## Solution

We have state methods to get what we need from a block.

This PR is a draft until it passed newly added CI tests.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

### CODE REVIEW

File: zebrad/tests/common/lightwalletd/proto/compact_formats.proto [hunk 1]
```diff
@@ -1,44 +1,55 @@
...
-    bytes header = 6;           // (hash, prevHash, and time) OR (full header)
-    repeated CompactTx vtx = 7; // zero or more compact transactions from this block
+    uint32 protoVersion = 1;         // the version of this wire format, for storage
+    uint64 height = 2;               // the height of this block
+    bytes hash = 3;                  // the ID (hash) of this block, same as in block explorers
+    bytes prevHash = 4;              // the ID (hash) of this block's predecessor
+    uint32 time = 5;                 // Unix epoch time when the block was mined
+    bytes header = 6;                // (hash, prevHash, and time) OR (full header)
+    repeated CompactTx vtx = 7;      // zero or more compact transactions from this block
+    ChainMetadata chainMetadata = 8; // information about the state of the chain as of this block
```
Comment (arya2) [inline zebrad/tests/common/lightwalletd/proto/compact_formats.proto:31]:
```suggestion
    ChainMetadata ChainMetadata = 8; // information about the state of the chain as of this block
```

Comment (oxarbitrage) [reply zebrad/tests/common/lightwalletd/proto/compact_formats.proto:31]: I copied the proto file from https://github.com/zcash/lightwalletd/blob/master/walletrpc/compact_formats.proto#L31

Comment (arya2) [reply zebrad/tests/common/lightwalletd/proto/compact_formats.proto:31]: Running full lightwalletd sync here: https://github.com/ZcashFoundation/zebra/actions/runs/5811606354

**Comment by arya2** [comment]
@Mergifyio update

File: zebrad/tests/common/lightwalletd/proto/compact_formats.proto (hunk 1 above)
Comment (arya2) [reply zebrad/tests/common/lightwalletd/proto/compact_formats.proto:31]: It failed, but may have cached the updated state for the block height checked in the test that's failing.

**Comment by oxarbitrage** [comment]
@Mergifyio update

**Comment by oxarbitrage** [comment]
I am still not sure what is going on with the CI here as it seems that it still using the old lightwalletd cache.

I tried running the test  of this PR locally and i was able to see the debug log i added [here](https://github.com/ZcashFoundation/zebra/blob/b58caca315fc63821ad8a24f2fddfa69cabecdd7/zebrad/tests/common/lightwalletd/wallet_grpc_test.rs#L169) and the test pass:

```
$ cargo test wallet_grpc_tests --features lightwalletd-grpc-tests -- --ignored --nocapture
...
2023-08-11T23:07:28.892747Z  INFO Some(ChainMetadata { sapling_commitment_tree_size: 1170439, orchard_commitment_tree_size: 2 })
...
test lightwalletd_wallet_grpc_tests ... ok
```

**Comment by oxarbitrage** [comment]
```
./zcash-rpc-diff 6666 getblock 2000000
```
getblock_diff1


```
./zcash-rpc-diff 6666 getblock 2000001
```
getblock_diff2


```
./zcash-rpc-diff 6666 getblock 2000002
```
getblock_diff3


```
./zcash-rpc-diff 6666 getblock 1890007
```
getblock_diff4


```
/zcash-rpc-diff 6666 getblock 1600000
```
getblock_diff5

**Review by arya2 — APPROVED** [review]
This looks great, thank you!

File: zebra-rpc/src/methods.rs [hunk 2]
```diff
@@ -1524,6 +1584,39 @@ impl GetRawTransaction {
...
+    #[serde(skip_serializing_if = "SaplingTrees::is_empty")]
+    sapling: SaplingTrees,
+    #[serde(skip_serializing_if = "OrchardTrees::is_empty")]
+    orchard: OrchardTrees,
+}
+
+/// Sapling note commitment tree information.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct SaplingTrees {
+    size: u64,
```
Comment (arya2) [inline zebra-rpc/src/methods.rs:1599]: Optional/Nitpick:

```suggestion
    /// position of the most recently appended Sapling leaf + 1
    size: u64,
```

File: zebra-rpc/src/methods.rs [hunk 3]
```diff
@@ -1524,6 +1584,39 @@ impl GetRawTransaction {
...
+impl SaplingTrees {
+    fn is_empty(&self) -> bool {
+        self.size == 0
+    }
+}
+
+/// Orchard note commitment tree information.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct OrchardTrees {
+    size: u64,
```
Comment (arya2) [inline zebra-rpc/src/methods.rs:1611]: Optional/Nitpick:

```suggestion
    /// position of the most recently appended Orchard leaf + 1
    size: u64,
```

**Comment by arya2** [comment]
@Mergifyio refresh

**Comment by gustavovalverde** [comment]
@merfigyio refresh

**Comment by gustavovalverde** [comment]
@mergifyio refresh

**Pull request merged**